### ISSUE TITLE
Lazyarrays in uproot4

### ISSUE THREAD

**Issue by tamasgal** [scikit-hep/uproot4#82]
The only references to lazy-loading (of arrays) I could find in the code so far are those using `uproot4.lazy()`, which requires a file and an object path.

Is there a plan to provide the old `.lazyarray()` interface as it was available in `uproot3`? Or does the current design make it hard to provide a lazyarray-interface for already opened files?

Btw. I noticed that the `.array()` method is very fast in many of our files which require custom interpretations (which is really nice!) but still, the memory requirement is sometimes huge (due to large dtypes) compared to what a user usually extracts from our branches and the Python GC is going crazy.

Also we usually deal with large amounts of branches and the overhead of the file opening might not be negligible.

**Comment by jpivarski** [comment]
> The only references to lazy-loading (of arrays) I could find in the code so far are those using uproot4.lazy(), which requires a file and an object path.

I am trying to cut down on the number of entry points, so there's only one way to make lazy arrays, `uproot4.lazy`. However, it accepts already opened TTree objects in its `files` argument (something that definitely needs to be documented). As such, we do have the ability to get lazy arrays from already open files, but the fundamental interface to lazy arrays is multi-file. (For example, you can make lazy arrays from a _list_ of already opened TTree objects.)

> Btw. I noticed that the .array() method is very fast in many of our files which require custom interpretations (which is really nice!) but still, the memory requirement is sometimes huge (due to large dtypes) compared to what a user usually extracts from our branches and the Python GC is going crazy.

If it's all in a single branch (single `AsDtype`), there's no way to read only part of it. If only a few parts are actually used by users, that's a problem with the file: the branch needs to be split. You'd have the same issue in ROOT. You can reduce the size of the lazily loaded chunks with `step_size`, which can in principle reduce your memory footprint, but if the problem is an over-active garbage collector, dividing the work into a larger number of smaller chunks would make the garbage collector busier.

I always recommend `iterate` over `lazy` in situations where memory needs to be controlled: lazy arrays would approach problems like

```python
arrays.branch1 + arrays.branch2 + arrays.branch3
```

by scanning through all chunks in `arrays.branch1` and `arrays.branch2` to produce `arrays.branch1 + arrays.branch2`, then scan through all chunks in that and `arrays.branch3` to produce the final sum. There's nowhere to insert the information that there's another `+` coming up and you'd really rather complete both sums in each chunk before moving on to the next chunk. The `iterate` method requires changes to user code:

```python
for arrays in uproot.iterate("files*.root", ["branch1", "branch2", "branch3"]):
    arrays.branch1 + arrays.branch2 + arrays.branch3
```

but those changes in user code are informative: now we know that everything inside the bounded loop should be completed before moving on to the next chunk. I've been looking at ways to use Dask to insert that information with lazy semantics (nothing happens until the user calls `.compute()`, which specifies the operations to be grouped per chunk in a different way), but I've been having some troubles with Dask's idea of what an array is (not jagged).

> Also we usually deal with large amounts of branches and the overhead of the file opening might not be negligible.

To make a lazy array, we somehow have to know the number of entries in all files and names/types of branches to expect (in their intersection, at least). @nsmith- is considering an alternative to the `lazy` function that gets this information some other way, and thus does not have to preemptively open all the files. That solution would require some sort of database, though, and the goal of `lazy` is for convenience in exploring files, something that wouldn't be the case if you had to set up a database first.

**Comment by tamasgal** [comment]
Alright, thanks for the explanation!

Regarding large dtypes, I though that you maybe do something fancy inside awkward or the memory mapping of the files to skip fields, but I have not thought much about it. Surely, iteration makes more sense to keep the memory footprint low!

Since we need to wrap `uproot` anyways in a class which creates a much more user-friend (and tab-completed) access to very long branch addresses, it's easy to maintain the old lazyarray functionality anyways, if still requested...

**Comment by jpivarski** [comment]
It's because the leaf values are interleaved in the TBasket buffer that gets compressed as a single unit. The whole thing has to be loaded to be decompressed. Then even selecting one field and letting the rest go out of scope still doesn't put the whole buffer out of scope because selecting a field is a view-slice in NumPy:, obtained by changing the strides, not by copying. So the original (all fields) becomes the `base` of the selected array.
